src/turtle/env/model/TestNode.d

```diff
@@ -0,0 +1,273 @@
+        port via restart().
+
+    ***************************************************************************/
+
+    final public void stop ( )
+    {
+        verify(this.state == State.Running, "Node is not running");
+
+        this.stopListeners(theScheduler.epoll);
+        this.shutdown();
+        this.unregister(); // Remove from turtle nodes registry
+        this.state = State.Stopped;
+    }
+
+    /***************************************************************************
+
+        Generate nodes files for the fake nodes. If the node supports the neo
+        protocol then the neo nodes file will also be written. The files are
+        named:
+        * this.id ~ ".nodes"
+        * this.id ~ ".neo.nodes"
+
+        Params:
+            directory = The directory the files will be written to.
+
+    ***************************************************************************/
+
+    public void genConfigFiles ( cstring directory )
+    {
+        shell("mkdir -p " ~ directory);
+
+        auto legacyfile = new File(directory ~ "/" ~ this.id ~ ".nodes",
+            File.WriteCreate);
+        scope (exit) legacyfile.close();
+
+        legacyfile.write(this.node_item.Address ~ ":" ~
+            Integer.toString(this.node_item.Port));
+        legacyfile.write("\n");
+
+        auto neofile = new File(directory ~ "/" ~ this.id ~ ".neo.nodes",
+            File.WriteCreate);
+        scope (exit) neofile.close();
+
+        neofile.write(this.node_item.Address ~ ":" ~
+            Integer.toString(this.neo_address.port));
+        neofile.write("\n");
+    }
+
+    /***************************************************************************
+
+        ITurtleEnv interface method implementation. Should not be called
+        manually.
+
+        Uses turtle env addition registry to stop tracking errors after all
+        tests have finished. This is necessary because applications don't do
+        clean connection shutdown when terminating, resulting in socket errors
+        being reported on node side.
+
+    ***************************************************************************/
+
+    public void unregister ( )
+    {
+        this.log_errors = false;
+    }
+
+    /***************************************************************************
+
+        Log errors, if logging is enabled.
+
+    ***************************************************************************/
+
+    private void onError ( Exception exception, IAdvancedSelectClient.Event,
+        IConnectionHandlerInfo )
+    {
+        if (!this.log_errors)
+            return;
+
+        log.warn("Ignoring exception: {} ({}:{})",
+            exception.message(), exception.file, exception.line);
+    }
+
+    /***************************************************************************
+
+        Removes all data from the fake node service.
+
+    ***************************************************************************/
+
+    abstract public void clear ( );
```

Comment:
Hm, what's the reason for having this abstract method?

Reply:
True, it's not required at this level.